### PR TITLE
Convert Tango SDF protobuf dumps; optional Python bindings

### PR DESCRIPTION
**Pull request by genemerewether**
Hi Helen and Marius - I restructured the Tango interface code to live in a separate ROS package, like we discussed. I'll hopefully have another PR (or add to this one - your choice) in another week or so for functionality involving known occupied or free space, and some more Python bindings.

- I changed some of the Layer and Block class members to `protected` so that I could access them in the derived Tango Interface classes.
- I added batch query functions to EsdfMap to reduce overhead of calling from Python
- Added some simple binaries for offline processing of TSDFs and ESDFs

### CODE REVIEW

**Comment by helenol** [comment]
add to whitelist

**Comment by genemerewether** [comment]
@helenol - would it be possible to get read-only access to the output of your Jenkins CI? Or just a copy paste of the output? Thanks :)

**Comment by helenol** [comment]
Hi! Sorry yes, didn't know you guys didn't have accesss - the issue is with the submodule.
```
First time build. Skipping changelog.
 > git remote # timeout=10
 > git submodule init # timeout=10
 > git submodule sync # timeout=10
 > git config --get remote.origin.url # timeout=10
 > git submodule init # timeout=10
 > git config -f .gitmodules --get-regexp ^submodule\.(.*)\.url # timeout=10
 > git config --get submodule.voxblox/pybind11.url # timeout=10
FATAL: Command "git config --get submodule.voxblox/pybind11.url" returned status code 1:
stdout: 
stderr: 
hudson.plugins.git.GitException: Command "git config --get submodule.voxblox/pybind11.url" returned status code 1:
stdout: 
stderr: 
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1903)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1871)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1867)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1512)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommand(CliGitAPIImpl.java:1524)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.getSubmoduleUrl(CliGitAPIImpl.java:1124)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$7.execute(CliGitAPIImpl.java:1050)
	at hudson.plugins.git.extensions.impl.SubmoduleOption.onCheckoutCompleted(SubmoduleOption.java:102)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1182)
	at hudson.scm.SCM.checkout(SCM.java:496)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1281)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:604)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.matrix.MatrixBuild.run(MatrixBuild.java:313)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Started calculate disk usage of build
```

I'll do the PR review later today and see if I can fix the submodule as well.

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[submodule "voxblox/pybind11"]
```
Comment (helenol) [inline .gitmodules]: Ah yeah I think this is the issue... Probably remove this one and keep the other.

File: voxblox_tango_interface/test/tango_tsdf_to_esdf.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  // Same number of voxels per side for ESDF as with TSDF
+  esdf_config.esdf_voxels_per_side = layer_from_file->voxels_per_side();
+  // Same voxel size for ESDF as with TSDF
+  esdf_config.esdf_voxel_size = layer_from_file->voxel_size();
+  esdf_map_.reset(new EsdfMap(esdf_config));
+  EsdfIntegrator::Config esdf_integrator_config;
+  // Make sure that this is the same as the truncation distance OR SMALLER!
+  esdf_integrator_config.min_distance_m = esdf_config.esdf_voxel_size;
+  // esdf_integrator_config.min_distance_m =
```
Comment (helenol) [inline voxblox_tango_interface/test/tango_tsdf_to_esdf.cc]: Please remove :)

**Comment by helenol** [comment]
And sorry for the delay with reviewing, and thanks for fixing the build!

File: voxblox/include/voxblox/io/layer_io.h [hunk 3]
```diff
@@ -137,6 +137,66 @@ bool LoadBlocksFromFile(
   return true;
 }
 
+/*TODO(mereweth@jpl.nasa.gov) - best name for this function?
+ * This function may be useful when debugging a malformed protobuf dump
+ * Also prevents segfaulting Python using file_path EsdfMap constructor
+ */
+template <typename VoxelType>
+typename Layer<VoxelType>::Ptr LoadOrCreateLayerHeader(
```
Comment (genemerewether) [inline voxblox/include/voxblox/io/layer_io.h]: Ok! My next PR will propose replacing this with a better-engineered overload for LoadLayer, still for Python bindings. In the meantime, I'll add a more descriptive comment and a note about the upcoming PR.

File: voxblox/src/core/esdf_map.cc [hunk 4]
```diff
@@ -50,4 +50,123 @@ bool EsdfMap::isObserved(const Eigen::Vector3d& position) const {
...
+        else {
+          did_all_fit = false;
+        }
+        count++;
+      }
+    }
+  }
+
+  if (!did_all_fit) {
+    throw std::runtime_error(std::string("Unable to store ") + std::to_string(count) + " values.");
```
Comment (genemerewether) [inline voxblox/src/core/esdf_map.cc]: Pybind11 converts std::exceptions into equivalent Python exceptions, which can be caught in Python. This seems pretty Pythonic, and I tried to only add exceptions where I was creating a new function specifically to bind in Python. Would that be OK to add exceptions in that specific case?

Does this function seem like something you'd like to use (aside from generating Python bindings?)

File: voxblox_ros/launch/cow_and_lady_dataset.launch [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <launch>
   <arg name="play_bag" default="true" />
-  <arg name="bag_file" default="/Users/helen/data/lady_cow_dataset/data.bag"/>
+  <arg name="bag_file" default="/Users/mereweth/snappy/tsdf_resources/cow_and_lady.bag"/>
```
Comment (genemerewether) [inline voxblox_ros/launch/cow_and_lady_dataset.launch]: hah! somehow that slipped back in. Will fix.

**Comment by genemerewether** [comment]
@helenol - would it be possible to merge this? Or are there any other changes desired?

Thanks,
Gene

**Comment by genemerewether** [comment]
@helenol - ping :)

**Comment by helenol** [comment]
@genemerewether sorry again, yes this is good to merge! Thanks!

**Comment by genemerewether** [comment]
ah! I don't have write access to be able to merge it - would you mind doing it? Thanks :)

**Pull request merged**